solutions/cses-2431.mdx

```diff
@@ -0,0 +1,169 @@
+---
+id: cses-2431
+source: CSES
+title: Digit Queries
+author: Nathan Gong
+---
+
+## Explanation
+
+Since $k$ can be as big as $10^{18}$, there is no way we can approach this
+problem through brute force. Instead, we'll need to work out an algorithm that
+can answer queries in logarithmic time.
+
+Let's group numbers by the amount of digits they have. An observation we can
+make is that for any $n \geq 1$, there are $9 \cdot 10^{n-1}$ numbers in the
+group with numbers of length $n$. Additionally, for each group, the amount of
+digits from all of the group's numbers combined is $n$ times the amount of
+numbers in the group. For instance:
+
+- Length $1$ numbers: $1$...$9$ -> $9$ numbers -> $1\cdot9=9$ digits total
+- Length $2$ numbers: $10...99$ -> $90$ numbers -> $2\cdot90=180$ digits total
+- Length $3$ numbers: $100...999$ -> $900$ numbers -> $3\cdot900=2700$ digits
+  total
+- And so on...
+
+To find which group the number that contains the $k$th digit is in, we can
+iterate through groups until the sum of the total number of digits from all the
+groups we've processed becomes greater than or equal to $k$. In other words, we
+continue increasing the number of groups we look at (starting from $1$-digit
+numbers) until the following is true:
+
+$$
+\sum^{\text{\# of groups}}_{n=1}{n\cdot9\cdot10^{n-1}\geq k}
+$$
+
+Once the condition becomes true, we know that the $k$th digit must fall into the
+last group used in the sum. From there, we can calculate the exact number that
+the $k$th digit is in, which we can then use to find the value of the $k$th
+digit with modular arithmetic.
```

Comment:
How exactly do you use modular arithmetic to do this? Also, modular arithmetic is a gold topic while this is a bronze problem, so ideally you should use something else (or explain it differently)

Reply:
I can add a brief explanation. This is really basic modular arithmetic (essentially just using k-1 % n to find the exact position of the number the digit is in) that I think is normal for bronze/silver.